### DISCUSSION TITLE
Problem accessing CEFSharp JavascriptObject in Popup

### DISCUSSION THREAD

**Discussion by aimcosoftware** [cefsharp/CefSharp#4459] [General]
When adding a `JavascriptObject` to a parent browser and popup, I can see the object in both.

In the parent I can call it's methods and properties, but not in the popup. I get the error `Object Not Found Matching Id:1`

Looking further: if I give the popup instance `JavascriptObject` a unique name, it works as expected.

I'm assuming this is expected behaviour. `JavascriptObjects` are global, referenced by name?

In order to have a `JavascriptObject` for each browser instance, it needs a unique name?

Thanks

**Comment by amaitland** [comment]
Bound objects are global to the **Renderer Process**. (They're stored in a static dictionary). Popups share the same Renderer Process as their parent. 

- https://developer.chrome.com/blog/inside-browser-part3/#inner-workings-of-a-renderer-process
- https://www.chromium.org/developers/design-documents/multi-process-architecture/#sharing-the-renderer-process

Main (non popup) browsers that share the same origin can be isolated using a [RequestContext](https://github.com/cefsharp/CefSharp/wiki/General-Usage#request-context-browser-isolation). 

> In the parent I can call it's methods and properties, but not in the popup. I get the error `Object Not Found Matching Id:1`

The information stored in the **Renderer Process** is basically just meta data describing the object and it's ID. Multiple browsers sharing the same **Renderer Process** share the same cache(static dictionary), so you end up with a different object ID even if the rest of the meta data is the same.

> In order to have a `JavascriptObject` for each browser instance, it needs a unique name?

If you are using a `ChromiumWebBrowser` instance to host the popup via the `newBrowser` param then yes a unique name (so the object gets a unique matching ID).

**Reply by aimcosoftware** [reply]
@amaitland Great, thanks for the detailed info. I do try to find the answer before posting, but never seem to find the results you link to ;-(

**Reply by amaitland** [reply]
> The information stored in the **Renderer Process** is basically just meta data describing the object and it's ID. Multiple browsers sharing the same **Renderer Process** share the same cache(static dictionary), so you end up with a different object ID even if the rest of the meta data is the same.

Just to clarify this is a `CefSharp` limitation, the `Chromium` architecture has changed since the original implementation. `JavaScript Objects` could be stored per browser  (#2306). It's a fiddly change and difficult to test so to date hasn't been actioned.

**Reply by aimcosoftware** [reply]
It's OK for me, I've just added the instance hash to the name, would it work to do something similar?
Then the architecture doesn't need to change, just the reference, which could be optional.

**Reply by amaitland** [reply]
> Just to clarify this is a `CefSharp` limitation, the `Chromium` architecture has changed since the original implementation. `JavaScript Objects` could be stored per browser (#2306). It's a fiddly change and difficult to test so to date hasn't been actioned.

A draft PR is now available #4475

Fairly extensive testing will be required before it's ready to merge.

**Comment by aimcosoftware** [comment]
I'm still having one issue with creating and returning `newBrowser` in `OnBeforePopup`.

I'm trying to work out how to allow a new renderer process.

I have tried creating a new `RequestContext`, but it doesn't seem to change anything.
If I don't call `Browser.SetAsPopup()`, as expected I get a new process, but other issues.

The below works exactly as expected, but in the parent process.
```
Browser = New ChromiumWebBrowser()
Browser.SetAsPopup()
newBrowser = Browser
windowInfo.SetAsChild(newBrowser.Handle)
Return False
```
Not sure if this just isn't possible, or a CEFSharp/CEF issue.
Any help appreciated, thanks.

**Reply by aimcosoftware** [reply]
Seems like this is by design: navigating to same site, uses the same renderer process. A new site, gets a new process.
